### PR TITLE
PlateauLR, a composable version of ReduceLROnPlateau (and extensible LRSchedule step API to support it)

### LINKED ISSUE #68978: SequentialLR cannot be used with ReduceLROnPlateau due to .step() not allowing for optional arguments
## 🐛 Bug

Currently, SequentialLR can only be used with schedulers that inherit from _LRScheduler or in other words adhere to the way .step() is called, i.e. without any arguments. This is not the case for the built in ReduceLROnPlateau as it takes in a metric value. Therefore, when calling SequentialLR.step() without arguments, ReduceLROnPlateau will raise an error once its milestone is reached. However, when calling SequentialLR.step(metric), SequentialLR will raise an error due to too many arguments.

Also see https://github.com/PyTorchLightning/pytorch-lightning/issues/10759

## To Reproduce
```python
scheduler1 = ConstantLR(self.opt, factor=0.1, total_iters=2)
scheduler2 = ReduceLROnPlateau (self.opt, gamma=0.9)
scheduler = SequentialLR(self.opt, schedulers=[scheduler1, scheduler2], milestones=[2])
for epoch in range(100):
   train(...)
   validate(...)
   scheduler.step()  # raises error after epoch>=2
   scheduler.step(metric)  # raises error immediately
```

## Expected behavior
SequentialLR should allow for arbitrary arguments in step to be used with the underlying scheduler. Either the user takes care of passing the right argument at the right time or there is a built in mechanic (if that is even possible).

### LINKED ISSUE #110761: Using ChainedScheduler with ReduceLROnPlateau leads to unexpected keyword argument error
### 🚀 The feature, motivation and pitch

I want to use ReduceLROnPlateau together with another schedule, say a warm-up or cosine schedule.
My understanding is that the suggested way to do that is through ``ChainedScheduler``.
However, when I try to call ``step`` with a validation loss, I get a ``TypeError``, since ``ChainedScheduler`` doesn't accept ``kwargs`` in ``step``.
I'm still pretty new to pytorch, so please let me know if I'm overlooking some other solution.

I don't see a super obvious solution. One would be to pass all ``kwargs`` to all steps in the chain, but that would mean all the schedulers would have to accept arbitrary ``kwargs`` which would hide errors. The other option that comes to mind would be to use inspect to figure out which parameters to pass to which scheduler. Or the ``step`` of ``ChainedScheduler`` could take a dict mapping from schedulers to parameters, but that seems more complicated than just calling all the schedules explicitly and not using ``ChainedScheduler``.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pupeno**
Fixes #68978  
Fixes #110761

Video overview of this PR to hopefully save you some time and hassle: https://youtu.be/X9ZAnNlwVII

Introduces `PlateauLR`, a composable `LRScheduler`-based replacement for `ReduceLROnPlateau`.

`PlateauLR` preserves the plateau-reduction behavior of `ReduceLROnPlateau`, but follows the normal `LRScheduler` lifecycle: it initializes through the base class, computes learning rates through `get_lr()`, and accepts the monitored value as a keyword-only `metrics` argument:

```python
scheduler = PlateauLR(optimizer)
scheduler.step(metrics=val_loss)
```

The change also:

- deprecates `ReduceLROnPlateau` with migration guidance to `PlateauLR`
- allows built-in schedulers and composition schedulers to accept an optional keyword-only `metrics` argument
- forwards metrics through `SequentialLR` and `ChainedScheduler`, including nested compositions and milestone handoffs
- preserves compatibility with third-party schedulers that retain the pre-`metrics` `step()` signature
- updates scheduler documentation and the scheduler illustration to use `PlateauLR`

`ReduceLROnPlateau` remains rejected by `SequentialLR` and `ChainedScheduler`, users needing composition should migrate to `PlateauLR`.

Sample code in here: https://github.com/pupeno/pytorch-workspace/tree/main/195634_plateau_lr_composable

An alternative approach is available in #195633. The two PRs are mutually exclusive.

This PR depends on https://github.com/pytorch/pytorch/pull/196242

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
I think the third-party scheduler compatibility check should also recognise `**kwargs`.

`_schedulers_accept_metrics` currently tests only whether the literal name `"metrics"` appears in `inspect.signature(type(scheduler).step).parameters`. A scheduler implemented as `def step(self, **kwargs)` legitimately accepts `metrics`, but will be classified as not accepting it and `SequentialLR` / `ChainedScheduler` will silently drop the metric.

Could this treat a `VAR_KEYWORD` parameter as metrics-capable as well? A small regression test with a third-party-style `step(**kwargs)` would cover the case.

**Comment by pupeno** [comment]
@sylvesterkaczmarek good point. At some point I implemented a kwargs pass-through so the API would be open to other keyword arguments being added in the future, and the composability would just work. I liked the outcome, but the code felt quite cumbersome. I think your point shows that we really need this. I'll make the change later today when I'm not working.

Do you think this PR has a chance? As a first time contributor, it felt like this is quite a big change I'm proposing.

**Comment by pupeno** [comment]
@sylvesterkaczmarek I applied a fix here that uses kwargs instead of metrics, so it's generic for future step() implementations that might need different arguments. Once we properly get rid of epoch and make it a stricter API, it would look cleaner.
